persistent/persistence.py

```diff
@@ -70,21 +70,24 @@ def _get_jar(self):
         return self.__jar
 
     def _set_jar(self, value):
-        if self.__jar is not None:
-            if self.__jar != value:
-                raise ValueError('Already assigned a data manager')
+        if self._p_is_in_cache() and value is not None and self.__jar != value:
+            # The C implementation only forbids changing the jar
+            # if we're already in a cache. Match its error message
+            raise ValueError('can not change _p_jar of cached object')
```

Comment:
It is in the C implementation :( https://github.com/zopefoundation/persistent/blob/master/persistent/cPersistence.c#L1091